### PR TITLE
test: fix unstable TestPurgeWorkerRelay

### LINKED ISSUE #1997: Quality reinforcement (13): fix unstable test cases
previous one https://github.com/pingcap/dm/issues/1790

- [ ] ha_master https://github.com/pingcap/dm/issues/1762
- [ ] ~testForEtcd.TestWorkerKeepAlive DEAD LOOP [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1293/pipeline)~
- [x] metrics & metrics causes [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1681/pipeline/) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1745/pipeline/142) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1759/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1773/pipeline) #2003
- [ ] ~syncer unit test [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1309/pipeline)~
- [x] server_test.go:406 testMaster.TestQueryStatus [jenkins](https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/dm_ghpr_new_test/runs/1510/nodes/140/steps/871/log/?start=0)
- [x] testMaster UT [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1554/pipeline) [jenkins] https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/2025/pipeline
- [ ] testElectionSuite.TestFailToStartLeader
```
[2021-08-10T03:55:24.792Z] election_test.go:91:
[2021-08-10T03:55:24.792Z]     c.Assert(leaderID, check.Equals, cfg1.Name)
[2021-08-10T03:55:24.792Z] ... obtained string = "dm-master-2"
[2021-08-10T03:55:24.792Z] ... expected string = "dm-master-1"
```
- [ ] testMaster.TestOfflineMember [jenkins](https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/dm_ghpr_new_test/runs/1696/nodes/140/steps/586/log/?start=0)
- [ ] ~sharding2 [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1754/pipeline)~
   runs too slow
- [ ] DM-UpdateBARule [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1757/pipeline)
- [x] shardddl1 DropAddColumn [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1799/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1821/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1837/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1858/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1865/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1854/pipeline/160) #2020 
- [ ] testReaderSuite.TestParseFilerelayLogUpdatedOrNewCreated [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1806/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1972/pipeline) #2040 
```
reader_test.go:296:
    c.Assert(err, IsNil)
... value *errors.withStack = decode relay meta toml file failed: open /tmp/check-2076299996949179932/4/b60868af-5a6f-11e9-9ea3-0242ac160006.000001/relay.meta: no such file or directory ("decode relay meta toml file failed: open /tmp/check-2076299996949179932/4/b60868af-5a6f-11e9-9ea3-0242ac160006.000001/relay.meta: no such file or directory")
```
- [x] testMaster.TestPurgeWorkerRelay [jenkins](https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/dm_ghpr_new_test/runs/1810/nodes/141/steps/901/log/?start=0) https://github.com/pingcap/dm/pull/2007
```
Current running test=testMaster.TestPurgeWorkerRelay
server_test.go:406:
    // TODO: why this will get context cancel?
    c.Assert(ha.KeepAlive(ctx, t.etcdTestCli, workerName, keepAliveTTL), check.IsNil)
... value *errors.errorString = &errors.errorString{s:"context canceled"} ("context canceled")
```
- [x] sequence_sharding_optimistic [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1876/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1917/pipeline/176) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1957/pipeline) #2037 
- [ ] dmtl pause-relay [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1877/pipeline)
- [x] metrics check_ticker_interval [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1879/pipeline/) #2022 
- [ ] master start too slowly [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1927/pipeline/) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1932/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1937/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1937/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1549/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1951/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1995/pipeline/)
- [x] github.com/pingcap/dm/dm/master/shardddl UT timeout [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1936/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1755/pipeline) #2041 
- [x] downstream_more_clolumn [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1944/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/2048/pipeline) #2056 
- [ ] testScheduler.TestScheduler [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1945/pipeline)
- [ ] ha_cases [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1954/pipeline)
- [x] shardddl2 DropAddColumn [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1968/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1982/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/2027/pipeline) #2052 
- [x] github.com/pingcap/dm/dm/master UT dead lock [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1970/pipeline) #2041 
- [ ] testEtcdUtilSuite.TestEtcdUtil [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1986/pipeline) [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/2127/pipeline)
- [x] ha_cases bind: address already in use [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1983/pipeline) #2044
- [ ] all_mode query-stauts takes too much time [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/1993/pipeline)
- [ ] handle_error_2 DM-4215 [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/2008/pipeline)
- [ ] import_goroutine_leak [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/2017/pipeline)
- [ ] ~~testMaster.TestCollectSourceConfigFilesV1Import [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/2074/pipeline)~~
```
[2021-08-27T06:47:27.170Z] bootstrap_test.go:97:
[2021-08-27T06:47:27.170Z]     c.Assert(cfgs[cfg1.SourceID], DeepEquals, cfg1)
[2021-08-27T06:47:27.170Z] ... obtained *config.SourceConfig = &config.SourceConfig{EnableGTID:false, AutoFixGTID:false, RelayDir:"relay-dir", MetaDir:"", Flavor:"mysql", Charset:"", EnableRelay:false, RelayBinLogName:"", RelayBinlogGTID:"", UUIDSuffix:0, SourceID:"mysql-replica-01", From:config.DBConfig{Host:"127.0.0.1", Port:3306, User:"root", Password:"123456", MaxAllowedPacket:(*int)(nil), Session:map[string]string{"time_zone":"+00:00"}, Security:(*config.Security)(nil), RawDBCfg:(*config.RawDBConfig)(nil)}, Purge:config.PurgeConfig{Interval:3600, Expires:0, RemainSpace:15}, Checker:config.CheckerConfig{CheckEnable:true, BackoffRollback:config.Duration{Duration:300000000000}, BackoffMax:config.Duration{Duration:300000000000}, CheckInterval:config.Duration{Duration:5000000000}, BackoffMin:config.Duration{Duration:1000000000}, BackoffJitter:true, BackoffFactor:2}, ServerID:0x65, Tracer:map[string]interface {}{}, CaseSensitive:true, Filters:[]*filter.BinlogEventRule{}} ("{\"enable-gtid\":false,\"auto-fix-gtid\":false,\"relay-dir\":\"relay-dir\",\"meta-dir\":\"\",\"flavor\":\"mysql\",\"charset\":\"\",\"enable-relay\":false,\"relay-binlog-name\":\"\",\"relay-binlog-gtid\":\"\",\"source-id\":\"mysql-replica-01\",\"from\":{\"host\":\"127.0.0.1\",\"port\":3306,\"user\":\"root\",\"max-allowed-packet\":null,\"session\":{\"time_zone\":\"+00:00\"},\"security\":null},\"purge\":{\"interval\":3600,\"expires\":0,\"remain-space\":15},\"checker\":{\"check-enable\":true,\"backoff-rollback\":{\"Duration\":\"5m0s\"},\"backoff-max\":{\"Duration\":\"5m0s\"}},\"server-id\":101,\"case-sensitive\":true,\"filters\":[]}")
[2021-08-27T06:47:27.171Z] ... expected *config.SourceConfig = &config.SourceConfig{EnableGTID:false, AutoFixGTID:false, RelayDir:"relay-dir", MetaDir:"", Flavor:"mysql", Charset:"", EnableRelay:false, RelayBinLogName:"", RelayBinlogGTID:"", UUIDSuffix:0, SourceID:"mysql-replica-01", From:config.DBConfig{Host:"127.0.0.1", Port:3306, User:"root", Password:"123456", MaxAllowedPacket:(*int)(nil), Session:map[string]string{"time_zone":"+00:00"}, Security:(*config.Security)(nil), RawDBCfg:(*config.RawDBConfig)(nil)}, Purge:config.PurgeConfig{Interval:3600, Expires:0, RemainSpace:15}, Checker:config.CheckerConfig{CheckEnable:true, BackoffRollback:config.Duration{Duration:300000000000}, BackoffMax:config.Duration{Duration:300000000000}, CheckInterval:config.Duration{Duration:5000000000}, BackoffMin:config.Duration{Duration:1000000000}, BackoffJitter:true, BackoffFactor:2}, ServerID:0x65, Tracer:map[string]interface {}{}, CaseSensitive:false, Filters:[]*filter.BinlogEventRule{}} ("{\"enable-gtid\":false,\"auto-fix-gtid\":false,\"relay-dir\":\"relay-dir\",\"meta-dir\":\"\",\"flavor\":\"mysql\",\"charset\":\"\",\"enable-relay\":false,\"relay-binlog-name\":\"\",\"relay-binlog-gtid\":\"\",\"source-id\":\"mysql-replica-01\",\"from\":{\"host\":\"127.0.0.1\",\"port\":3306,\"user\":\"root\",\"max-allowed-packet\":null,\"session\":{\"time_zone\":\"+00:00\"},\"security\":null},\"purge\":{\"interval\":3600,\"expires\":0,\"remain-space\":15},\"checker\":{\"check-enable\":true,\"backoff-rollback\":{\"Duration\":\"5m0s\"},\"backoff-max\":{\"Duration\":\"5m0s\"}},\"server-id\":101,\"case-sensitive\":false,\"filters\":[]}")
```
- [ ] syncer.go WaitGroup data race 
   https://github.com/pingcap/dm/blob/e008ef3e2d068da5c27e8df12700927d0f8719f7/syncer/syncer.go#L669
   and inside `s.Run` there're many `s.wg.Add`

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
close #1997

### What is changed and how it works?
only when ha.KeepAlive successful put the keepalive KV, it will return no error when exit. so we wait it to put.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Review by Ehco1996 — APPROVED** [review]

File: dm/master/server_test.go [hunk 1]
```diff
@@ -194,7 +195,6 @@ func (t *testMaster) TearDownSuite(c *check.C) {
 }
 
 func (t *testMaster) SetUpTest(c *check.C) {
-	c.Logf("Current running test=%s", c.TestName())
```
Comment (lichunzhu) [inline dm/master/server_test.go:197]: Why delete this?

**Review by lichunzhu — APPROVED** [review]
Rest LGTM

File: dm/master/server_test.go (hunk 1 above)
Comment (lance6716) [reply dm/master/server_test.go:197]: go test already printed this

**Pull request merged**